File: frontend/src/types/code.ts
/** 代码解析的出入参类型（spec §6.2 code 行 + §8.4，与后端 StaticReport 结构对齐）。 */

/**
 * 可运行语言（spec §3.1 假设 3：演示环境具备 Python 与 Node 运行时）。
 * 定义在 types 而非 api —— api 层反过来依赖 types，反过来引会造成循环。
 */
export type CodeLanguage = 'python' | 'javascript'

export interface LineStats {
  total: number
  code: number
  blank: number
  comment: number
}

export interface FunctionInfo {
  name: string
  line: number
  args: number
  complexity: number
}

export interface ClassInfo {
  name: string
  line: number
  methods: number
}

export interface ComplexitySummary {
  max: number
  average: number
  worst: string | null
}

export interface UnusedVariable {
  name: string
  line: number
}

export interface BareExcept {
  line: number
}

export interface SyntaxIssue {
  line: number
  message: string
}

export interface StaticReport {
  language: 'python' | 'javascript'
  lines: LineStats
  functions: FunctionInfo[]
  classes: ClassInfo[]
  complexity: ComplexitySummary
  issues: { unused_variables: UnusedVariable[]; bare_excepts: BareExcept[] }
  syntax_error: SyntaxIssue | null
}

export interface AIReport {
  content: string
  provider: string
  model: string
  token_usage: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  } | null
  usage_estimated: boolean
  degraded: boolean
  fallback_reason: string | null
}

export interface CodeAnalysisOut {
  analysis_id: string
  language: string
  static_report: StaticReport
  ai_report: AIReport | null
  reused: boolean
}

// ---------------------------------------------------------------- 代码运行（P4 / spec §8.3）

/** spec §5 的封闭取值。 */
export type RunStatus =
  | 'accepted'
  | 'runtime_error'
  | 'timeout'
  | 'memory_exceeded'
  | 'blocked'

/**
 * 四层资源限制的实测留痕（不是配置回显）。
 * `applied` 为 false 表示该层在当前平台没设上 —— 此时 degraded_layers 会带上它。
 */
export interface LimitDetail {
  executed: boolean
  error: string | null
  blacklist: { rule: string | null; executed: boolean } | null
  wall_clock: { limit_s: number; elapsed_s: number; triggered: boolean }
  cpu: { limit_s: number; applied: boolean; error: string | null; triggered: boolean }
  memory: {
    limit_bytes: number
    sampled: boolean
    interval_ms: number
    peak_bytes: number
    samples: number
    triggered: boolean
  }
  file_size: { limit_bytes: number; applied: boolean; error: string | null }
  output: {
    limit_bytes: number
    stdout_bytes: number
    stderr_bytes: number
    stdout_truncated: boolean
    stderr_truncated: boolean
  }
  process_group: { killed: boolean }
  degraded_layers: string[]
  platform: string
}

export interface CodeRunOut {
  run_id: string
  status: RunStatus
  stdout: string
  stderr: string
  exit_code: number | null
  duration_ms: number
  limit_detail: LimitDetail | null
}

export interface CodeSessionOut {
  id: string
  language: CodeLanguage
  title: string
  source_code: string
  updated_at: string
}

export interface CodeRunHistoryItem {
  id: string
  language: CodeLanguage
  source_code: string
  status: RunStatus
  stdout: string
  stderr: string
  exit_code: number | null
  duration_ms: number
  limit_detail: LimitDetail | null
  created_at: string
}

export interface Paged<T> {
  items: T[]
  total: number
}

File: backend/app/schemas/code.py
"""code 端点出入参（spec §6.2 code 行）。"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.domain.code.analysis import ANALYSIS_LANGUAGES
from app.domain.code.execution import RUN_LANGUAGES

# 源码上限沿用 /code/analyze（spec §8.4 未另定，两处保持一致）
SOURCE_MAX = 20000
# stdin 不落 LLM、也不进哈希，只喂给子进程；给一个比源码宽松得多的上限
STDIN_MAX = 200000


class CodeAnalyzeIn(BaseModel):
    language: str = Field(pattern=f"^({'|'.join(ANALYSIS_LANGUAGES)})$")
    source: str = Field(min_length=1, max_length=SOURCE_MAX)


class CodeAnalysisOut(BaseModel):
    """spec §6.2 的最小集为 {static_report, ai_report, analysis_id}；
    `reused` 是增量字段 —— 前端据此提示「与上次分析相同，已复用历史结果」。"""

    model_config = ConfigDict(from_attributes=True)

    analysis_id: str
    language: str
    static_report: dict
    ai_report: dict | None = None
    reused: bool = False


# ---------------------------------------------------------------- 代码运行（P4）

class CodeRunIn(BaseModel):
    """spec §6.2：`POST /code/run {language, source, stdin}`。"""

    language: str = Field(pattern=f"^({'|'.join(RUN_LANGUAGES)})$")
    source: str = Field(min_length=1, max_length=SOURCE_MAX)
    stdin: str = Field(default="", max_length=STDIN_MAX)


class CodeRunOut(BaseModel):
    """spec §6.2 的七字段。`run_id` 指向落库的 `CodeRun` 行。"""

    model_config = ConfigDict(from_attributes=True)

    run_id: str
    status: str
    stdout: str
    stderr: str
    exit_code: int | None = None
    duration_ms: int
    limit_detail: dict | None = None


class CodeSessionIn(BaseModel):
    language: str = Field(pattern=f"^({'|'.join(RUN_LANGUAGES)})$")
    title: str = Field(default="未命名会话", max_length=100)
    source_code: str = Field(default="", max_length=SOURCE_MAX)


class CodeSessionPatch(BaseModel):
    """三个字段都可独立更新；`None` 表示「不改」，与「改成空串」区分。"""

    language: str | None = Field(default=None, pattern=f"^({'|'.join(RUN_LANGUAGES)})$")
    title: str | None = Field(default=None, max_length=100)
    source_code: str | None = Field(default=None, max_length=SOURCE_MAX)


class CodeSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    language: str
    title: str
    source_code: str
    updated_at: datetime


class CodeRunHistoryItem(BaseModel):
    """历史列表不含 stdout 全文以外的东西，但**含** limit_detail —— 学生回看
    「上次为什么被杀」需要它。"""

    model_config = ConfigDict(from_attributes=True)

    id: str
    language: str
    source_code: str
    status: str
    stdout: str
    stderr: str
    exit_code: int | None = None
    duration_ms: int
    limit_detail: dict | None = None
    created_at: datetime
